cdap-app-templates/cdap-etl/cdap-etl-core/src/test/java/io/cdap/cdap/etl/common/record/StructuredRecordComparatorTest.java

```diff
@@ -0,0 +1,412 @@
+    }
+
+    testSingleFieldNotEqual(Schema.nullableOf(INT_SCHEMA), 0, null);
+    testSingleFieldNotEqual(Schema.nullableOf(LONG_SCHEMA), 0L, null);
+    testSingleFieldNotEqual(Schema.nullableOf(FLOAT_SCHEMA), 0f, null);
+    testSingleFieldNotEqual(Schema.nullableOf(DOUBLE_SCHEMA), 0d, null);
+    testSingleFieldNotEqual(Schema.nullableOf(STRING_SCHEMA), "0", null);
+    testSingleFieldNotEqual(Schema.nullableOf(BYTES_SCHEMA), new byte[] { 0 }, null);
+    testSingleFieldNotEqual(Schema.nullableOf(BOOL_SCHEMA), false, null);
+  }
+
+  @Test
+  public void testMultipleSimpleFields() {
+    Schema schema = Schema.recordOf("x",
+                                    Schema.Field.of("int", INT_SCHEMA),
+                                    Schema.Field.of("long", LONG_SCHEMA),
+                                    Schema.Field.of("float", FLOAT_SCHEMA),
+                                    Schema.Field.of("double", DOUBLE_SCHEMA),
+                                    Schema.Field.of("string", STRING_SCHEMA),
+                                    Schema.Field.of("bool", BOOL_SCHEMA),
+                                    Schema.Field.of("bytes", BYTES_SCHEMA));
+
+    StructuredRecord r1 = StructuredRecord.builder(schema)
+      .set("int", 0)
+      .set("long", 0L)
+      .set("float", 0f)
+      .set("double", 0d)
+      .set("string", "0")
+      .set("bool", false)
+      .set("bytes", new byte[] { 0 })
+      .build();
+    Assert.assertEquals(0, COMPARATOR.compare(r1, r1));
+    StructuredRecord r2 = copy(r1).set("int", 1).build();
+    testRecordsNotEqual(r1, r2);
+
+    r2 = copy(r1).set("long", 1L).build();
+    testRecordsNotEqual(r1, r2);
+
+    r2 = copy(r1).set("float", 1f).build();
+    testRecordsNotEqual(r1, r2);
+
+    r2 = copy(r1).set("double", 1d).build();
+    testRecordsNotEqual(r1, r2);
+
+    r2 = copy(r1).set("string", "1").build();
+    testRecordsNotEqual(r1, r2);
+
+    r2 = copy(r1).set("bool", true).build();
+    testRecordsNotEqual(r1, r2);
+
+    r2 = copy(r1).set("bytes", new byte[] { 0, 0 }).build();
+    testRecordsNotEqual(r1, r2);
+  }
+
+  @Test
+  public void testSimpleArrays() {
+    Schema schema = Schema.recordOf("x", Schema.Field.of("x", Schema.arrayOf(INT_SCHEMA)));
+
+    // test with lists
+    StructuredRecord r1 = StructuredRecord.builder(schema).set("x", Arrays.asList(0, 1, 2, 3, 4, 5)).build();
+    StructuredRecord r2 = StructuredRecord.builder(schema).set("x", Arrays.asList(0, 1, 2, 3, 4)).build();
+    Assert.assertEquals(0, COMPARATOR.compare(r1, r1));
+    testRecordsNotEqual(r1, r2);
+
+    // test comparing list with array
+    r2 = StructuredRecord.builder(schema).set("x", new int[] { 0, 1, 2, 3, 4, 5 }).build();
+    Assert.assertEquals(0, COMPARATOR.compare(r1, r2));
+    r2 = StructuredRecord.builder(schema).set("x", new int[] { 0, 1, 2, 3, 4 }).build();
+    testRecordsNotEqual(r1, r2);
+
+    r2 = StructuredRecord.builder(schema).set("x", Arrays.asList(5, 4, 3, 2, 1, 0)).build();
```

Comment:
I was wondering why this passes, given the hasNext() bug above. I think if you use `{5, 4, 3, 2, 1, 0}` here, it would fail and return 0. Because you'd iterate over a collection and an array, the collection will return hasNext() = true but the array will return false, the loop breaks and it returns 0.

Reply:
fixed